src/gallium/drivers/lima/lima_texture.c

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2018 Lima Project
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sub license,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice (including the
+ * next paragraph) shall be included in all copies or substantial portions
+ * of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#include "util/u_memory.h"
+#include "util/u_upload_mgr.h"
+#include "util/u_math.h"
+#include "util/u_debug.h"
+#include "util/u_transfer.h"
+
+#include "lima_bo.h"
+#include "lima_context.h"
+#include "lima_texture.h"
+#include "lima_resource.h"
+
+static uint32_t pipe_format_to_lima(enum pipe_format pformat)
+{
+   unsigned swap_chans, flag1, format;
+
+   switch (pformat) {
+   case PIPE_FORMAT_R8G8B8A8_UNORM:
+   case PIPE_FORMAT_R8G8B8X8_UNORM:
+      swap_chans = 1;
+      flag1 = 0;
+      format = LIMA_TEXEL_FORMAT_RGBA_8888;
+      break;
+   case PIPE_FORMAT_B8G8R8A8_UNORM:
+   case PIPE_FORMAT_B8G8R8X8_UNORM:
+      swap_chans = 0;
+      flag1 = 0;
+      format = LIMA_TEXEL_FORMAT_RGBA_8888;
+      break;
+   case PIPE_FORMAT_R8G8B8_UNORM:
+      swap_chans = 1;
+      flag1 = 0;
+      format = LIMA_TEXEL_FORMAT_RGB_888;
+      break;
+   case PIPE_FORMAT_B5G6R5_UNORM:
+      swap_chans = 0;
+      flag1 = 0;
+      format = LIMA_TEXEL_FORMAT_BGR_565;
+      break;
+   default:
+      assert(0);
+      break;
+   }
+
+   return (swap_chans << 7) | (flag1 << 6) | format;
+}
+
+static void
+lima_update_tex_desc(struct lima_context *ctx, struct lima_sampler_state *sampler,
+                     struct lima_sampler_view *texture, void *pdesc)
+{
+   uint32_t *desc = pdesc;
+   unsigned width, height, layout;
+   struct pipe_resource *prsc = texture->base.texture;
+   struct lima_resource *lima_res = lima_resource(prsc);
+
+   width = prsc->width0;
```

Comment:
How about the stride if width0 != stride?

Reply:
I'm not sure whether hardware supports it - at least I couldn't find support for stride in limare and I don't know how to specify custom stride with GLES 2.0 - glPixelStorei() doesn't support GL_UNPACK_ROW_LENGTH for GLES 2.0, see https://www.khronos.org/registry/OpenGL-Refpages/es2.0/xhtml/glPixelStorei.xml

Reply:
I'm OK if you want to put handling width0 != stride case latter if it needs more investigation. There're two problems need to be figure out:
1. can texture sampler hardware handle texture width/height not aligned to some value like render target (aligned to 16)
2. there will be some texture used as render target (FBO), so it will be aligned to 16 (width0 != stride case), how to handle this

Reply:
Good point, I haven't tested corner cases, including texture width/height not aligned to some value.